Allow sending ephemeral data to application services (#2018)
matrix-org/matrix-spec: content/application-service-api.md

{{% http-api spec="application-service" api="transactions" %}}

> ##### Pushing EDUs

Reviewer: I know that the MSC used the term "EDU", but I think it's somewhat misleading. In particular: we do *not* forward `m.direct_to_device` EDUs to ASes (not least because such EDUs contain multiple to-device messages, possibly for other users).

Also: presence/receipts/typing from users on the same server don't arrive via EDUs, so the current wording would imply that the AS would only receive updates from users on other servers, which I don't believe was the intention of the MSC.

Also also: the event format used by the C-S API (and hence, now, by the A-S API) is quite different from that used by EDUs in the S-S API.

Could we, instead of saying "EDU", say "ephemeral data" ?

{{% added-in v="1.13" %}}

If the `receive_ephemeral` settings is enabled in the [registration](#registration)

> file, homeservers MUST send all EDUs that are relevant to the application

Reviewer: not all EDUs; device list updates and to-device messages are excluded.

```suggestion
file, homeservers MUST send ephemeral data that is relevant to the application
```

service via the transaction API, using the `ephemeral` property of the request's
body. This property is an array that is effectively a combination of the
`presence` and `ephemeral` sections of the client-server [`/sync`](/client-server-api/#get_matrixclientv3sync)

> API. It means that there are only three event types that can currently occur:
> [`m.presence`](/client-server-api/#mpresence), [`m.typing`](/client-server-api/#mtyping),
> and [`m.receipt`](/client-server-api/#mreceipt).

Reviewer: ```suggestion
API. 

There are currently three event types that can be delivered to an application service:
[`m.presence`](/client-server-api/#mpresence), [`m.typing`](/client-server-api/#mtyping),
and [`m.receipt`](/client-server-api/#mreceipt).
```

EDUs that can be associated to a particular room (i.e. `m.typing` or `m.receipt`)

> MUST also include a `room_id` property to identify the room that they were sent

Reviewer: ```suggestion
MUST use the same format as the client-server API, with the addition of a `room_id` property at the top level to identify the room that they were sent
```

in. They should be sent to the application service under the same rules as
regular events, meaning that the application service must have registered
interest in the room itself, or in a user that is in the room.
[Private read receipts](/client-server-api/#private-read-receipts) should only
be sent for users matching one of the application service's namespaces. Normal
read receipts and threaded read receipts are always sent.

EDUs which are not associated with a particular room (i.e. `m.presence`), should
be sent to the application service if the EDU would apply contextually. For
example, a presence update for a user an application service shares a room with,

> or matching one of the application service's namespaces.

Reviewer: It might be easier to describe each of the three event types separately, with a bullet list, rather than trying to describe the behaviour via generic rules.
